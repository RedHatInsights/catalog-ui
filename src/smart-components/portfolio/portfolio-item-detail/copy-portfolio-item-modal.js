import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';
import {
  FormGroup,
  Modal,
  TextContent,
  Text,
  TextVariants
} from '@patternfly/react-core';
import { componentTypes } from '@data-driven-forms/react-form-renderer';

import FormRenderer from '../../common/form-renderer';
import { getPortfolioItemApi } from '../../../helpers/shared/user-login';
import {
  copyPortfolioItem,
  fetchPortfolioItemsWithPortfolio
} from '../../../redux/actions/portfolio-actions';
import asyncFormValidator from '../../../utilities/async-form-validator';
import {
  listPortfolios,
  getPortfolio
} from '../../../helpers/portfolio/portfolio-helper';
import { PORTFOLIO_ITEM_ROUTE } from '../../../constants/routes';

const loadPortfolios = (filter) =>
  listPortfolios(filter, { limit: 100, offset: 0 }).then(({ data }) =>
    data.map(({ name, id }) => ({ value: id, label: name }))
  );

const copySchema = (portfolioName, portfolioChange, nameFetching) => ({
  fields: [
    {
      component: 'value-only',
      name: 'portfolio_item_name',
      label: 'Name',
      value: portfolioName
    },
    {
      component: componentTypes.SELECT,
      name: 'portfolio_id',
      label: 'Portfolio',
      isRequired: true,
      loadOptions: asyncFormValidator(loadPortfolios),
      onChange: portfolioChange,
      isDisabled: nameFetching,
      isSearchable: true
    }
  ]
});

const ValueOnly = ({ name, label, value }) => (
  <FormGroup label={label} fieldId={name}>
    <TextContent>
      <Text component={TextVariants.h6}>{value}</Text>
    </TextContent>
  </FormGroup>
);

ValueOnly.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired
};

const CopyPortfolioItemModal = ({
  portfolioId,
  portfolioItemId,
  closeUrl,
  search
}) => {
  const dispatch = useDispatch();
  const { push } = useHistory();
  const [submitting, setSubmitting] = useState(false);
  const [name, setName] = useState();
  const [nameFetching, setNameFetching] = useState(false);

  useEffect(() => {
    getPortfolioItemApi()
      .getPortfolioItemNextName(portfolioItemId, portfolioId)
      .then(({ next_name }) => setName(next_name));
  }, []);
  const onSubmit = async (values) => {
    setSubmitting(true);
    const portfolio = await getPortfolio(values.portfolio_id);
    dispatch(copyPortfolioItem(portfolioItemId, values, portfolio))
      .then(({ id, service_offering_source_ref }) =>
        push({
          pathname: PORTFOLIO_ITEM_ROUTE,
          search: `?portfolio=${values.portfolio_id}&portfolio-item=${id}&source=${service_offering_source_ref}`
        })
      )
      .then(
        () =>
          values.portfolio_id === portfolioId &&
          dispatch(fetchPortfolioItemsWithPortfolio(portfolioId))
      )
      .catch(() => setSubmitting(false));
  };

  const portfolioChange = (portfolioId) => {
    setNameFetching(true);
    return getPortfolioItemApi()
      .getPortfolioItemNextName(portfolioItemId, portfolioId)
      .then(({ next_name }) => {
        setName(next_name);
      })
      .then(() => setNameFetching(false));
  };

  return (
    <Modal
      isOpen
      title="Copy product"
      onClose={() =>
        push({
          pathname: closeUrl,
          search
        })
      }
      isSmall
    >
      <FormRenderer
        initialValues={{ portfolio_id: portfolioId, portfolio_item_name: name }}
        schema={copySchema(name, portfolioChange, nameFetching)}
        onSubmit={onSubmit}
        onCancel={() =>
          push({
            pathname: closeUrl,
            search
          })
        }
        formContainer="modal"
        componentMapper={{ 'value-only': ValueOnly }}
        buttonsLabels={{ submitLabel: 'Save' }}
        disableSubmit={submitting ? ['pristine', 'diry'] : []}
      />
    </Modal>
  );
};

CopyPortfolioItemModal.propTypes = {
  closeUrl: PropTypes.string.isRequired,
  portfolioId: PropTypes.string,
  portfolioItemId: PropTypes.string.isRequired,
  search: PropTypes.string.isRequired
};
export default CopyPortfolioItemModal;
