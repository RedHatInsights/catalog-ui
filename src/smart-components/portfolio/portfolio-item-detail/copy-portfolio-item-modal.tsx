/* eslint-disable react/prop-types */
import React from 'react';
import { useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';
import { Modal } from '@patternfly/react-core';

import FormRenderer from '../../common/form-renderer';
import { getPortfolioItemApi } from '../../../helpers/shared/user-login';
import {
  copyPortfolioItem,
  fetchPortfolioItemsWithPortfolio,
  fetchSelectedPortfolio
} from '../../../redux/actions/portfolio-actions';
import {
  copyPortfolioItem as copyPortfolioItemS,
  fetchPortfolioItemsWithPortfolio as fetchPortfolioItemsWithPortfolioS,
  fetchSelectedPortfolio as fetchSelectedPortfolioS
} from '../../../redux/actions/portfolio-actions-s';

import asyncFormValidator from '../../../utilities/async-form-validator';
import { listPortfolios } from '../../../helpers/portfolio/portfolio-helper';
import { PORTFOLIO_ITEM_ROUTE } from '../../../constants/routes';
import actionMessages from '../../../messages/actions.messages';
import labelMessages from '../../../messages/labels.messages';
import portfolioMessages from '../../../messages/portfolio.messages';
import useFormatMessage from '../../../utilities/use-format-message';
import {
  FormatMessage,
  Full,
  SelectOptions
} from '../../../types/common-types';
import {
  Portfolio,
  PortfolioItem,
  PortfolioItemNextName
} from '@redhat-cloud-services/catalog-client';

const loadPortfolios = (name: string) =>
  listPortfolios({ name }, { limit: 100, offset: 0 }).then(({ data }) =>
    data
      .filter(
        ({
          metadata: {
            user_capabilities: { update }
          }
        }) => update
      )
      .map(({ name, id }) => ({ value: id, label: name }))
  );

const copySchema = (
  getName: (value: string) => Promise<string | undefined>,
  formatMessage: FormatMessage,
  initialOptions: SelectOptions
) => ({
  fields: [
    {
      component: 'copy-name-display',
      name: 'portfolio_item_name',
      label: formatMessage(labelMessages.name),
      getName,
      fieldSpy: 'portfolio_id'
    },
    {
      component: 'initial-select',
      name: 'portfolio_id',
      label: formatMessage(labelMessages.portfolio),
      isRequired: true,
      loadOptions: asyncFormValidator(loadPortfolios),
      isSearchable: true,
      options: initialOptions,
      menuIsPortal: true
    }
  ]
});

export interface CopyPortfolioItemModalProps {
  closeUrl: string;
  portfolioId?: string;
  portfolioItemId: string;
  search?: string;
  portfolioName?: string;
}

const CopyPortfolioItemModal: React.ComponentType<CopyPortfolioItemModalProps> = ({
  portfolioId,
  portfolioItemId,
  closeUrl,
  search,
  portfolioName
}) => {
  const formatMessage = useFormatMessage();
  const dispatch = useDispatch();
  const { push } = useHistory();

  const onSubmit = async (values: Full<PortfolioItem>) => {
    /**
     * dispatch redux action to set selected portfolio in store
     * this will ensure that correct portfolio data will be loaded after the redirect occurs
     */
    const { value: portfolio } = await dispatch(
      fetchSelectedPortfolio(values.portfolio_id) as Promise<{
        value: Full<Portfolio>;
      }>
    );
    return dispatch(
      (copyPortfolioItem(
        portfolioItemId,
        values,
        portfolio
      ) as unknown) as Promise<PortfolioItem>
    )
      .then(({ id, service_offering_source_ref }) =>
        push({
          pathname: PORTFOLIO_ITEM_ROUTE,
          search: `?portfolio=${values.portfolio_id}&portfolio-item=${id}&source=${service_offering_source_ref}`
        })
      )
      .then(
        () =>
          values.portfolio_id === portfolioId &&
          dispatch(
            window.catalog?.standalone
              ? fetchPortfolioItemsWithPortfolioS(portfolioId)
              : fetchPortfolioItemsWithPortfolio(portfolioId)
          )
      );
  };

  const portfolioChange = (portfolioId: string) =>
    (getPortfolioItemApi().getPortfolioItemNextName(
      portfolioItemId,
      portfolioId
    ) as Promise<PortfolioItemNextName>).then(({ next_name }) => next_name);

  return (
    <Modal
      isOpen
      title={formatMessage(portfolioMessages.copyItemTitle) as string}
      onClose={() =>
        push({
          pathname: closeUrl,
          search
        })
      }
      variant="small"
    >
      <FormRenderer
        initialValues={{ portfolio_id: portfolioId }}
        schema={copySchema(portfolioChange, formatMessage, [
          { value: portfolioId, label: portfolioName }
        ])}
        onSubmit={onSubmit}
        onCancel={() =>
          push({
            pathname: closeUrl,
            search
          })
        }
        templateProps={{
          submitLabel: formatMessage(actionMessages.save),
          disableSubmit: ['validating', 'submitting']
        }}
      />
    </Modal>
  );
};

export default CopyPortfolioItemModal;
