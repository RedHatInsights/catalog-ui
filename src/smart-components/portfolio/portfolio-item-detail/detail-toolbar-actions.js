import PropTypes from 'prop-types';
import React, { Fragment } from 'react';
import {
  Dropdown,
  DropdownItem,
  DropdownPosition,
  KebabToggle,
  LevelItem
} from '@patternfly/react-core';
import ButtonWithSpinner from '../../../presentational-components/shared/button-with-spinner';
import CatalogLink from '../../common/catalog-link';

const DetailToolbarActions = ({
  copyUrl,
  orderUrl,
  editUrl,
  workflowUrl,
  editSurveyUrl,
  isOpen,
  setOpen,
  isFetching,
  availability,
  userCapabilities: { update, copy }
}) => {
  const dropdownItems = [];
  if (update) {
    dropdownItems.push(
      <DropdownItem
        aria-label="Edit Portfolio"
        key="edit-portfolio-item"
        id="edit-portfolio-item"
        component={
          <CatalogLink pathname={editUrl} preserveSearch>
            Edit
          </CatalogLink>
        }
        role="link"
      />
    );
  }

  if (copy) {
    dropdownItems.push(
      <DropdownItem
        aria-label="Copy Portfolio"
        key="copy-portfolio-item"
        id="copy-portfolio-item"
        component={
          <CatalogLink pathname={copyUrl} preserveSearch>
            Copy
          </CatalogLink>
        }
        role="link"
      />
    );
  }

  return (
    <Fragment>
      <LevelItem>
        <CatalogLink
          isDisabled={isFetching || availability === 'unavailable'}
          pathname={orderUrl}
          preserveSearch
        >
          <ButtonWithSpinner
            isDisabled={isFetching || availability === 'unavailable'}
            showSpinner={isFetching}
            variant="primary"
            id="order-portfolio-item"
          >
            Order
          </ButtonWithSpinner>
        </CatalogLink>
      </LevelItem>
      <LevelItem style={{ marginLeft: 16 }}>
        <Dropdown
          isPlain
          onToggle={setOpen}
          onSelect={() => setOpen(false)}
          position={DropdownPosition.right}
          toggle={
            <KebabToggle
              id="portfolio-item-actions-toggle"
              onToggle={(isOpen) => setOpen(isOpen)}
            />
          }
          isOpen={isOpen}
          dropdownItems={[
            ...dropdownItems,
            <DropdownItem
              aria-label="Set approval"
              key="edit-approval_workflow"
              component={
                <CatalogLink pathname={workflowUrl} preserveSearch>
                  Set approval
                </CatalogLink>
              }
              role="link"
            />,
            <DropdownItem
              aria-label="Edit survey"
              key="edit-survey"
              component={
                <CatalogLink pathname={editSurveyUrl} preserveSearch>
                  Edit survey
                </CatalogLink>
              }
              role="link"
            />
          ]}
        />
      </LevelItem>
    </Fragment>
  );
};

DetailToolbarActions.propTypes = {
  orderUrl: PropTypes.string.isRequired,
  editUrl: PropTypes.string.isRequired,
  copyUrl: PropTypes.string.isRequired,
  editSurveyUrl: PropTypes.string.isRequired,
  workflowUrl: PropTypes.string.isRequired,
  isOpen: PropTypes.bool,
  setOpen: PropTypes.func.isRequired,
  isFetching: PropTypes.bool,
  availability: PropTypes.oneOf(['available', 'unavailable']).isRequired,
  userCapabilities: PropTypes.shape({
    update: PropTypes.bool,
    copy: PropTypes.bool
  }).isRequired
};

DetailToolbarActions.defaultProps = {
  isFetching: false
};

export default DetailToolbarActions;
