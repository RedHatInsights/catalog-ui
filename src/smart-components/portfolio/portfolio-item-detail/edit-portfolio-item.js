import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { useLocation, useHistory } from 'react-router-dom';

import FormRenderer from '../../common/form-renderer';
import editPortfolioItemSchema from '../../../forms/edit-portfolio-item-form.schema';
import { updatePortfolioItem } from '../../../redux/actions/portfolio-actions';
import { Stack, StackItem } from '@patternfly/react-core';
import IconUpload from './icon-upload';
import { CATALOG_API_BASE } from '../../../utilities/constants';
import CardIcon from '../../../presentational-components/shared/card-icon';
import { addNotification } from '@redhat-cloud-services/frontend-components-notifications/cjs/actions';
import portfolioMessages from '../../../messages/portfolio.messages';
import useFormatMessage from '../../../utilities/use-format-message';

const EditPortfolioItem = ({
  cancelUrl,
  uploadIcon,
  resetIcon,
  product: { owner, created_at, updated_at, ...product },
  userCapabilities
}) => {
  const dispatch = useDispatch();
  const { push } = useHistory();
  const { search } = useLocation();
  const [image, setImage] = useState();
  const formatMessage = useFormatMessage();

  const submitChanges = (values) => {
    push({
      pathname: cancelUrl,
      search
    });
    uploadIcon(image).catch((error) => {
      dispatch(
        addNotification({
          variant: 'danger',
          title: formatMessage(portfolioMessages.portfolioItemIconTitle),
          description: error.data.errors[0].detail,
          dismissable: true
        })
      );
      setImage(undefined);
    });
    return dispatch(
      updatePortfolioItem({
        ...values,
        metadata: { user_capabilities: userCapabilities }
      })
    );
  };

  return (
    <Stack hasGutter>
      <StackItem key={product.icon_id || 'default'}>
        <IconUpload
          initialImage={image}
          setImage={setImage}
          uploadIcon={uploadIcon}
          resetIcon={resetIcon}
          enableReset={!!product.icon_id}
        >
          <CardIcon
            src={
              image ||
              `${CATALOG_API_BASE}/portfolio_items/${
                product.id
              }/icon?cache_id=${product.icon_id || 'default'}`
            } // we need ho add the query to prevent the browser caching when reseting the image
            sourceId={product.service_offering_source_ref}
            height={64}
          />
        </IconUpload>
      </StackItem>
      <StackItem>
        <FormRenderer
          initialValues={{ ...product }}
          onSubmit={(values) => submitChanges(values)}
          schema={editPortfolioItemSchema}
          templateProps={{
            disableSubmit: ['pristine', 'image_pristine']
          }}
          onCancel={() =>
            push({
              pathname: cancelUrl,
              search
            })
          }
        />
      </StackItem>
    </Stack>
  );
};

EditPortfolioItem.propTypes = {
  cancelUrl: PropTypes.string.isRequired,
  product: PropTypes.object.isRequired,
  userCapabilities: PropTypes.object.isRequired,
  uploadIcon: PropTypes.func.isRequired,
  resetIcon: PropTypes.func.isRequired
};

export default EditPortfolioItem;
