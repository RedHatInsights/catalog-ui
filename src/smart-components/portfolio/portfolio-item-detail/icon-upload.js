import React, { useState, useRef } from 'react';
import { PencilAltIcon } from '@patternfly/react-icons';
import {
  Spinner,
  Dropdown,
  DropdownToggle,
  DropdownItem
} from '@patternfly/react-core';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { addNotification } from '@redhat-cloud-services/frontend-components-notifications/redux';
import styled from 'styled-components';
import useFormatMessage from '../../../utilities/use-format-message';
import iconMessages from '../../../messages/icon.messages';

const UploadButton = styled.span`
  border: none;
  position: absolute;
  top: 0;
  left: 0;
  padding-left: 0;
  width: 100%;
  height: 100%;
  background-color: transparent;
  &::after {
    content: ' ';
    position: absolute;
    display: block;
    top: -0.3em;
    right: -0.3em;
    width: 1.5em;
    height: 1.5em;
    border-radius: 50%;
    background-color: rgba(255, 255, 255, 0.8);
    z-index: 0;
  }
  svg,
  .pf-c-spinner {
    z-index: 1;
    position: absolute;
    top: 0;
    right: 0;
  }
`;

const UploadIconWrapper = styled.div`
  position: relative;
  display: inline-block;
`;

const ImagePreview = styled.img`
  height: 64px;
  max-width: 300px;
  object-fit: cover;
`;

const StyledDropdown = styled(Dropdown)`
  .pf-c-dropdown__menu {
    top: 24px;
    left: calc(100% - 24px);
  }
`;

const IconUpload = ({ image, setImage, resetIcon, enableReset, children }) => {
  const formatMessage = useFormatMessage();
  const inputRef = useRef();
  const [isUploading, setIsUploading] = useState(false);
  const [isOpen, setIsOpen] = useState(false);
  const handleClick = () => {
    setIsOpen(false);
    return inputRef.current.click();
  };

  const handleReset = () => {
    setImage(undefined);
    setIsUploading(true);
    return resetIcon().then(() => setIsUploading(false));
  };

  const dropdownItems = [
    <DropdownItem onClick={handleClick} key="change-icon">
      {formatMessage(iconMessages.changeIcon)}
    </DropdownItem>,
    <DropdownItem
      isDisabled={!enableReset}
      onClick={handleReset}
      key="reset-icon"
    >
      {formatMessage(iconMessages.resetIcon)}
    </DropdownItem>
  ];

  return (
    <UploadIconWrapper>
      <input
        accept=".png,.svg,.jpeg"
        onChange={(event) => {
          setImage(event.target.files[0]);
        }}
        ref={inputRef}
        type="file"
        id="icon-upload"
        hidden
      />
      <StyledDropdown
        onSelect={() => setIsOpen(false)}
        isOpen={isOpen}
        isPlain
        disabled={isUploading}
        dropdownItems={dropdownItems}
        toggle={
          <DropdownToggle
            disabled={isUploading}
            toggleIndicator={null}
            onToggle={(isOpen, event) => {
              event.stopPropagation();
              setIsOpen(isOpen);
            }}
          >
            <UploadButton ouiaId={'upload-icon'}>
              {isUploading ? (
                <Spinner size="md" />
              ) : (
                <PencilAltIcon size="sm" />
              )}
            </UploadButton>
            {!image && children}
            {image && (
              <ImagePreview
                style={{ height: 64 }}
                src={URL.createObjectURL(image)}
                id={image.name}
              />
            )}
          </DropdownToggle>
        }
      />
    </UploadIconWrapper>
  );
};

IconUpload.propTypes = {
  uploadIcon: PropTypes.func.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node)
  ]),
  resetIcon: PropTypes.func.isRequired,
  enableReset: PropTypes.bool
};

export default IconUpload;
