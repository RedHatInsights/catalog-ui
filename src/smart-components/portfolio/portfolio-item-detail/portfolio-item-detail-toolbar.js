import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Route, Link } from 'react-router-dom';
import {
  Level,
  LevelItem,
  Text,
  TextContent,
  TextVariants,
  Button,
  Flex,
  Dropdown,
  KebabToggle,
  DropdownItem
} from '@patternfly/react-core';

import DetailToolbarActions from './detail-toolbar-actions';
import { CATALOG_API_BASE } from '../../../utilities/constants';
import CardIcon from '../../../presentational-components/shared/card-icon';
import TopToolbar from '../../../presentational-components/shared/top-toolbar';
import IconUpload from './icon-upload';
import ButtonWithSpinner from '../../../presentational-components/shared/button-with-spinner';
import { StyledLevelItem } from '../../../presentational-components/styled-components/level';

const PortfolioItemIconItem = ({ uploadIcon, id, sourceId }) => (
  <IconUpload uploadIcon={uploadIcon}>
    <CardIcon
      src={`${CATALOG_API_BASE}/portfolio_items/${id}/icon`}
      sourceId={sourceId}
      height={64}
    />
  </IconUpload>
);

PortfolioItemIconItem.propTypes = {
  uploadIcon: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired,
  sourceId: PropTypes.string.isRequired
};

export const PortfolioItemDetailToolbar = ({
  url,
  isOpen,
  product,
  setOpen,
  isFetching,
  uploadIcon,
  availability,
  userCapabilities
}) => (
  <TopToolbar breadcrumbsSpacing={false}>
    <Level>
      <StyledLevelItem alignEnd className="pf-l-flex">
        <PortfolioItemIconItem
          uploadIcon={uploadIcon}
          id={product.id}
          sourceId={product.service_offering_source_ref}
        />
        <TextContent className="pf-u-ml-md">
          <Text component={TextVariants.h1}>{product.name}</Text>
        </TextContent>
      </StyledLevelItem>
      <LevelItem style={{ minHeight: 36 }}>
        <Level>
          <Route
            exact
            path={url}
            render={(...args) => (
              <DetailToolbarActions
                isOpen={isOpen}
                setOpen={(open) => setOpen(open)}
                orderUrl={`${url}/order`}
                editUrl={`${url}/edit`}
                copyUrl={`${url}/copy`}
                editSurveyUrl={`${url}/edit-survey`}
                workflowUrl={`${url}/edit-workflow`}
                isFetching={isFetching}
                availability={availability}
                userCapabilities={userCapabilities}
                {...args}
              />
            )}
          />
        </Level>
      </LevelItem>
    </Level>
    <Level>
      <LevelItem>
        <TextContent>
          <Text component={TextVariants.small}>{product.distributor}</Text>
        </TextContent>
      </LevelItem>
    </Level>
  </TopToolbar>
);

PortfolioItemDetailToolbar.propTypes = {
  url: PropTypes.string.isRequired,
  isOpen: PropTypes.bool,
  product: PropTypes.shape({
    distributor: PropTypes.string,
    name: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired,
    service_offering_source_ref: PropTypes.string.isRequired
  }).isRequired,
  setOpen: PropTypes.func.isRequired,
  isFetching: PropTypes.bool,
  uploadIcon: PropTypes.func.isRequired,
  availability: PropTypes.oneOf(['available', 'unavailable']).isRequired
};

PortfolioItemDetailToolbar.defaultProps = {
  isFetching: false
};

const SurveyEditorDropdown = ({ handleResetSurvey }) => {
  const [isOpen, setIsOpen] = useState(false);
  return (
    <Dropdown
      isOpen={isOpen}
      isPlain
      onSelect={() => setIsOpen(false)}
      position="right"
      toggle={<KebabToggle onToggle={(isOpen) => setIsOpen(isOpen)} />}
      dropdownItems={[
        <DropdownItem
          onClick={handleResetSurvey}
          component="button"
          key="synchronize"
        >
          Restore to Ansible Tower version
        </DropdownItem>
      ]}
    />
  );
};

SurveyEditorDropdown.propTypes = {
  handleResetSurvey: PropTypes.func.isRequired
};

export const SurveyEditingToolbar = ({
  uploadIcon,
  product,
  handleSaveSurvey,
  closeUrl,
  search,
  isFetching,
  isValid,
  modified,
  handleResetSurvey
}) => (
  <TopToolbar breadcrumbsSpacing={false} breadcrumbs={true}>
    <Level>
      <StyledLevelItem alignEnd className="pf-l-flex">
        <PortfolioItemIconItem
          uploadIcon={uploadIcon}
          id={product.id}
          sourceId={product.service_offering_source_ref}
        />
        <TextContent className="pf-u-ml-md">
          <Text component={TextVariants.h1}>
            Editing survey: {product.name}
          </Text>
        </TextContent>
      </StyledLevelItem>
      <LevelItem>
        <Flex className="align-items-center">
          <ButtonWithSpinner
            variant="primary"
            showSpinner={isFetching}
            isDisabled={isFetching || !isValid}
            onClick={handleSaveSurvey}
          >
            Save
          </ButtonWithSpinner>
          <Link
            to={{
              pathname: closeUrl,
              search
            }}
          >
            <Button variant="link">Cancel</Button>
          </Link>
          {modified && (
            <SurveyEditorDropdown handleResetSurvey={handleResetSurvey} />
          )}
        </Flex>
      </LevelItem>
    </Level>
  </TopToolbar>
);

SurveyEditingToolbar.propTypes = {
  uploadIcon: PropTypes.func.isRequired,
  product: PropTypes.object,
  handleSaveSurvey: PropTypes.func.isRequired,
  closeUrl: PropTypes.string.isRequired,
  search: PropTypes.string.isRequired,
  isFetching: PropTypes.bool,
  isValid: PropTypes.bool,
  modified: PropTypes.bool,
  handleResetSurvey: PropTypes.func,
  userCapabilities: PropTypes.object.isRequired
};
