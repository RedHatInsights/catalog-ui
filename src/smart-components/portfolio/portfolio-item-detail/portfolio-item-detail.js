import React, { useEffect, useState, Fragment, lazy, Suspense } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Route, Switch, useRouteMatch } from 'react-router-dom';
import { Grid, GridItem, Alert } from '@patternfly/react-core';
import { Section } from '@redhat-cloud-services/frontend-components/components/cjs/Section';

import ItemDetailInfoBar from './item-detail-info-bar';
import ItemDetailDescription from './item-detail-description';
import { PortfolioItemDetailToolbar } from './portfolio-item-detail-toolbar';
import TopToolbar from '../../../presentational-components/shared/top-toolbar';
import { getPortfolioItemDetail } from '../../../redux/actions/portfolio-actions';
import {
  ProductLoaderPlaceholder,
  AppPlaceholder
} from '../../../presentational-components/shared/loader-placeholders';
import { uploadPortfolioItemIcon } from '../../../helpers/portfolio/portfolio-helper';
import useQuery from '../../../utilities/use-query';
import { PORTFOLIO_ITEM_ROUTE } from '../../../constants/routes';
import CatalogRoute from '../../../routing/catalog-route';

const SurveyEditor = lazy(() =>
  import(
    /* webpackChunkName: "survey-editor" */ '../../survey-editing/survey-editor'
  )
);

const requiredParams = ['portfolio', 'source', 'portfolio-item'];

const PortfolioItemDetail = () => {
  const [isOpen, setOpen] = useState(false);
  const [isFetching, setIsFetching] = useState(true);
  const dispatch = useDispatch();
  const [queryValues, search] = useQuery(requiredParams);
  const { url } = useRouteMatch(PORTFOLIO_ITEM_ROUTE);
  const {
    portfolioItem: {
      metadata: { user_capabilities: userCapabilities },
      ...portfolioItem
    },
    source
  } = useSelector(({ portfolioReducer: { portfolioItem } }) => portfolioItem);
  const portfolio = useSelector(
    ({ portfolioReducer: { selectedPortfolio } }) => selectedPortfolio
  );

  useEffect(() => {
    setIsFetching(true);
    dispatch(
      getPortfolioItemDetail({
        portfolioItem: queryValues['portfolio-item'],
        ...queryValues
      })
    )
      .then(() => setIsFetching(false))
      .catch(() => setIsFetching(false));
  }, [queryValues['portfolio-item']]);

  if (isFetching || Object.keys(portfolioItem).length === 0) {
    return (
      <Section className="global-primary-background full-height">
        <TopToolbar>
          <ProductLoaderPlaceholder />
        </TopToolbar>
      </Section>
    );
  }

  const availability = source.availability_status || 'unavailable';
  const unavailable = [source]
    .filter(({ notFound }) => notFound)
    .map(({ object }) => (
      <Alert
        className="pf-u-mb-sm"
        key={object}
        variant="warning"
        isInline
        title={`The ${object} for this product is no longer available`}
      />
    ));
  const uploadIcon = (file) => uploadPortfolioItemIcon(portfolioItem.id, file);
  return (
    <Fragment>
      <Switch>
        <CatalogRoute
          requiredCapabilities="update"
          userCapabilities={userCapabilities}
          path={`${url}/edit-survey`}
        >
          <Suspense fallback={<AppPlaceholder />}>
            <SurveyEditor
              closeUrl={url}
              search={search}
              uploadIcon={uploadIcon}
              portfolioItem={portfolioItem}
              portfolio={portfolio}
            />
          </Suspense>
        </CatalogRoute>
        <Route>
          <Section className="full-height global-primary-background">
            <PortfolioItemDetailToolbar
              uploadIcon={uploadIcon}
              url={url}
              isOpen={isOpen}
              product={portfolioItem}
              setOpen={setOpen}
              isFetching={isFetching}
              availability={availability}
              userCapabilities={userCapabilities}
            />
            {unavailable.length > 0 && (
              <div className="pf-u-mr-lg pf-u-ml-lg">{unavailable}</div>
            )}
            {source.availability_status === 'unavailable' && (
              <Alert
                className="pf-u-ml-lg pf-u-mr-lg"
                id="unavailable-alert-info"
                variant="info"
                isInline
                title="The platform for this product is unavailable"
              />
            )}
            <Grid hasGutter className="pf-u-p-lg">
              <GridItem md={3} lg={2}>
                <ItemDetailInfoBar
                  product={portfolioItem}
                  portfolio={portfolio}
                  source={source}
                />
              </GridItem>
              <GridItem md={9} lg={10}>
                <ItemDetailDescription
                  product={portfolioItem}
                  userCapabilities={userCapabilities}
                  url={url}
                  search={search}
                />
              </GridItem>
            </Grid>
          </Section>
        </Route>
      </Switch>
    </Fragment>
  );
};

export default PortfolioItemDetail;
