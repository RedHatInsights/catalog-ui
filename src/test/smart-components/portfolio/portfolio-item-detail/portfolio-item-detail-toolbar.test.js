import React from 'react';
import thunk from 'redux-thunk';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { shallowToJson } from 'enzyme-to-json';
import { MemoryRouter, Route } from 'react-router-dom';
import promiseMiddleware from 'redux-promise-middleware';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/';

import DetailToolbarActions from '../../../../smart-components/portfolio/portfolio-item-detail/detail-toolbar-actions';
import { PortfolioItemDetailToolbar } from '../../../../smart-components/portfolio/portfolio-item-detail/portfolio-item-detail-toolbar';

describe('<PortfolioItemDetailToolbar />', () => {
  let initialProps;
  let initialState;
  const middlewares = [thunk, promiseMiddleware, notificationsMiddleware()];
  let mockStore;

  const ComponentWrapper = ({ store, children, initialEntries = [] }) => (
    <Provider store={store}>
      <MemoryRouter initialEntries={initialEntries}>{children}</MemoryRouter>
    </Provider>
  );

  beforeEach(() => {
    initialProps = {
      url: '/foo',
      isOpen: false,
      product: {
        name: 'bar'
      },
      setOpen: jest.fn(),
      setWorkflow: jest.fn(),
      handleUpdate: jest.fn(),
      userCapabilities: {
        copy: true,
        update: true
      }
    };

    initialState = {
      breadcrumbsReducer: { fragments: [] },
      platformReducer: { platforms: [], platformIconMapping: {} }
    };
    mockStore = configureStore(middlewares);
  });

  it('should render correctly', () => {
    const wrapper = shallow(
      <MemoryRouter>
        <PortfolioItemDetailToolbar {...initialProps} />
      </MemoryRouter>
    ).find(PortfolioItemDetailToolbar);
    expect(
      shallowToJson(wrapper.find(PortfolioItemDetailToolbar))
    ).toMatchSnapshot();
  });

  it('should render view variant', () => {
    const store = mockStore(initialState);
    const wrapper = mount(
      <ComponentWrapper store={store} initialEntries={['/foo']}>
        <Route
          path="/foo"
          render={() => <PortfolioItemDetailToolbar {...initialProps} />}
        />
      </ComponentWrapper>
    );
    expect(wrapper.find(DetailToolbarActions)).toHaveLength(1);
  });
});
