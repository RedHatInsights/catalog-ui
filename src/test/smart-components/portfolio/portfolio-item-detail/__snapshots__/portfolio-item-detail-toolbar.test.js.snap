// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PortfolioItemDetailToolbar /> should render correctly 1`] = `
<PortfolioItemDetailToolbar
  handleUpdate={[MockFunction]}
  isFetching={false}
  isOpen={false}
  product={
    Object {
      "name": "bar",
    }
  }
  setOpen={[MockFunction]}
  setWorkflow={[MockFunction]}
  url="/foo"
  userCapabilities={
    Object {
      "copy": true,
      "update": true,
    }
  }
/>
`;
